src/main/java/edu/umich/ctools/esb/utils/WAPI.java

```diff
@@ -330,52 +353,51 @@ public WAPIResultWrapper reportError(int status) {
 		return new WAPIResultWrapper(status, errMsg, new JSONObject("{error : " + ERROR_MSG + "}"));
 	}
 
-	// For the next few methods there must be some nice way to pass in the method and
-	// get around repeating code but that is something to do in the future.
-
-	public WAPIResultWrapper getRequest(String request){
-		M_log.info("getRequest() called with request: " + request);
-		return getRequest(request,null);
-	}
-
-	//Make a get request, if error returned then try to renew token and ask again.
-	public WAPIResultWrapper getRequest(String request,HashMap<String,String> headers){
+	// Wrapper to deal with token renewal and query retry.
+	protected WAPIResultWrapper getOrPutRequestWithRetry(String requestType, String request,HashMap<String,String> headers){
 		WAPIResultWrapper wrappedResult = doGetRequest(request,headers);
-		M_log.info("getRequest() called with request: " + request+" headers: "+headers);
+		M_log.info("getOrPutRequestWithRetry() called with request type: " + requestType + " request: "+request+" headers: "+headers);
 
 		if(wrappedResult.getStatus()==HTTP_UNAUTHORIZED){
 			wrappedResult = renewToken();
 			if(wrappedResult.getStatus()==HTTP_SUCCESS){
-				wrappedResult = doGetRequest(request,headers);
+				headers = updateAuthHeader(headers);
+				if (GET.equals(requestType)) {
+					wrappedResult = doGetRequest(request,headers);
+				}
+				if (PUT.equals(requestType)) {
+					wrappedResult = doPutRequest(request,headers);
+				}
 			}
 		}
 		return wrappedResult;
 	}
 
-	// Make a request, if error returned then try to renew token and ask again.
-	public WAPIResultWrapper putRequest(String request){
-		return putRequest(request,null);
+
+	// Set auth header to given value.
+	private HashMap<String, String> updateAuthHeader(HashMap<String, String> headers,String token) {
+		headers.put(AUTHORIZATION, token);
+		return headers;
 	}
 
-	// Make a request, if error returned then try to renew token and ask again.
-	public WAPIResultWrapper putRequest(String request,HashMap<String,String> headers){
-		WAPIResultWrapper wrappedResult = doPutRequest(request,headers);
-		M_log.info("putRequest() called with request: " + request+" headers: "+headers);
+	// Set auth header to current token.
+	private HashMap<String, String> updateAuthHeader(HashMap<String, String> headers) {
+		String incoming_token = headers.get(AUTHORIZATION);
 
-		if(wrappedResult.getStatus()==HTTP_UNAUTHORIZED){
-			wrappedResult = renewToken();
-			if(wrappedResult.getStatus()==HTTP_SUCCESS){
-				wrappedResult = doPutRequest(request,headers);
-			}
+		// check for invalid situation.
+		if (incoming_token == this.token) {
+			M_log.info("updateAuthHeader: token did not change ???: "+incoming_token);
```

Comment:
should this be warning/error instead? Anyway, the token seems to be renewed nevertheless.